### PR TITLE
PR User Reviews

### PR DESCRIPTION
**Pull request by Lyubomy-r**
- in the User class, the added fields rating and fields receivedReviews, leftReviews to bidirectional connection.
- the rating field has been added in the User ProfileDTO class
- u.rating was added in the UserRepository class, in the Query method findUserFullInfoById()
- in the UserControllerTest class, the rating value was added in the shouldReturnMailResetPasswordResponseWithStatusCode200 test

-classes have been created ReviewController, UserReview, UserReviewDTO, UserReviewRepository, UserReviewResponseDTO, UserReviewSaveDTO, UserReviewService, UserReviewServiceImp 
-created tests ReviewControllerTest, UserReviewRepositoryTest, UserReviewServiceImpTest,

-INSERT INTO user_review was added in the data.sql file and rating was added to the previous INSERT into users

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -87,11 +86,11 @@
...
-        <dependency>
-            <groupId>org.postgresql</groupId>
-            <artifactId>postgresql</artifactId>
-            <scope>runtime</scope>
-        </dependency>
+<!--        <dependency>-->
+<!--            <groupId>org.postgresql</groupId>-->
+<!--            <artifactId>postgresql</artifactId>-->
+<!--            <scope>runtime</scope>-->
+<!--        </dependency>-->
```
Comment (MinoUni) [inline pom.xml]: Use non-in-memory db in runtime

File: pom.xml [hunk 2]
```diff
@@ -108,7 +107,7 @@
         <dependency>
             <groupId>com.h2database</groupId>
             <artifactId>h2</artifactId>
-            <scope>test</scope>
+            <scope>runtime</scope>
```
Comment (MinoUni) [inline pom.xml]: Return test scope

File: src/main/java/com/teamchallenge/bookti/review/ReviewController.java [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+                                            schema = @Schema(implementation = ErrorResponse.class))
+                            })
+            })
+    @GetMapping("/user/{userId}/received")
+    public ResponseEntity<List<UserReview>> findAllUserReceivedReviewsById(@PathVariable("userId") Integer userId) {
+
+        List<UserReview> response = userReviewService.findAllUserReceivedReviewsById(userId);
+        log.info("From ReviewController method findAllUserReviewById - /reviews/user/{userId} - return user reviews. ");
+        return ResponseEntity.status(HttpStatus.OK).body(response);
+    }
```
Comment (MinoUni) [inline src/main/java/com/teamchallenge/bookti/review/ReviewController.java]: Path userId with ReqParam
expl: /reviews?userId=<value> || /reviews?[ filter ] || in payload

File: src/main/java/com/teamchallenge/bookti/review/ReviewController.java [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+                    @ApiResponse(
+                            responseCode = "500",
+                            description = "INTERNAL_SERVER_ERROR",
+                            content = {
+                                    @Content(
+                                            mediaType = APPLICATION_JSON_VALUE,
+                                            schema = @Schema(implementation = ErrorResponse.class))
+                            })
+            })
+    @PostMapping("/user")
```
Comment (MinoUni) [inline src/main/java/com/teamchallenge/bookti/review/ReviewController.java]:
```suggestion
    @PostMapping
```
You work with review resources, add user data if it needed to payload or reqParam

File: src/main/java/com/teamchallenge/bookti/review/ReviewController.java [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+                            responseCode = "500",
+                            description = "INTERNAL_SERVER_ERROR",
+                            content = {
+                                    @Content(
+                                            mediaType = APPLICATION_JSON_VALUE,
+                                            schema = @Schema(implementation = ErrorResponse.class))
+                            })
+            })
+    @PreAuthorize("isAuthenticated() and hasRole('ROLE_USER')")
+    @DeleteMapping("/user/{reviewId}")
```
Comment (MinoUni) [inline src/main/java/com/teamchallenge/bookti/review/ReviewController.java]:
```suggestion
    @DeleteMapping("/{reviewId}")
```
As previous

File: src/main/java/com/teamchallenge/bookti/review/UserReviewDTO.java [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+        return UserReviewDTO.builder()
+                .id(user.getId())
+                .reviewerName(user.getReviewerName())
+                .message(user.getMessage())
+                .rating(user.getRating())
+                .creationDate(user.getCreationDate())
+                .avatarUrl(user.getAvatarUrl())
+                .owner(userProfile)
+                .build();
+    }
```
Comment (MinoUni) [inline src/main/java/com/teamchallenge/bookti/review/UserReviewDTO.java]: Replace with mapstruct mapper

File: src/main/java/com/teamchallenge/bookti/review/UserReviewServiceImp.java [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+    public List<UserReview> findAllUserReceivedReviewsById(Integer userId) {
+        if (userId > 0) {
+            List<UserReview> userReviewList = userReviewRepository.findAllUserReceivedReviewsById(userId);
+            log.info("From UserReviewServiceImp method - findAllUserReviewById - return list or empty list, received review to user: {}.", userId);
+
+            return userReviewList;
+        } else {
+            throw new UserNotFoundException("Write correct Users id.");
+        }
+    }
```
Comment (MinoUni) [inline src/main/java/com/teamchallenge/bookti/review/UserReviewServiceImp.java]:
```suggestion
    @Override
    public List<UserReview> findAllUserReceivedReviewsById(Integer userId) {
      if (!userRepository.existsById(userId)) {
        log.info("message");
        throw new UserNotFoundException(String.format(UserConstant.NOT_FOUND_MESSAGE, userId));        
      }
      List<UserReview> userReviewList = userReviewRepository.findAllUserReceivedReviewsById(userId);
      log.info("From UserReviewServiceImp method - findAllUserReviewById - return list or empty list, received review to user: {}.", userId);
      return userReviewList;
    }
```
If you need only check the existence of the model without reading the data -> use existsBy();
If you need to read model data and verify that the model is present -> use findBy();

File: src/main/java/com/teamchallenge/bookti/review/UserReviewServiceImp.java [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+    @Override
+    public List<UserReview> findAllUserLeftReviewById(Integer userId) {
+        if (userId > 0) {
+            List<UserReview> userReviewList = userReviewRepository.findAllUserLeftReviewById(userId);
+            log.info("From UserReviewServiceImp method - findAllUserReviewById - return list or empty list, Left reviews to user: {}.", userId);
+
+            return userReviewList;
+        } else {
+            throw new UserNotFoundException("Write correct Users id.");
+        }
```
Comment (MinoUni) [inline src/main/java/com/teamchallenge/bookti/review/UserReviewServiceImp.java]: same as previous

File: src/main/java/com/teamchallenge/bookti/review/UserReviewServiceImp.java [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+    private static UserReview createUserReview(UserReviewSaveDTO userReview, User userReviewer, User userOwner) {
+        return UserReview.builder()
+                .reviewerName(userReviewer.getFullName())
+                .avatarUrl(userReviewer.getAvatarUrl())
+                .message(userReview.getMessage())
+                .rating(userReview.getRating().setScale(1, RoundingMode.HALF_UP))
+                .owner(userOwner)
+                .reviewers(userReviewer)
+                .build();
+    }
```
Comment (MinoUni) [inline src/main/java/com/teamchallenge/bookti/review/UserReviewServiceImp.java]: replace with mapper using mapstruct

File: src/main/java/com/teamchallenge/bookti/review/UserReviewServiceImp.java [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+        } else {
+            throw new UserNotFoundException("Write correct Users id.");
+        }
+    }
+
+    @Transactional
+    @Override
+    public UserReviewResponseDTO save(UserReviewSaveDTO review, Integer reviewerId) {
+
+        if (checkValidUserId(reviewerId, review.getOwnerId())) {
```
Comment (MinoUni) [inline src/main/java/com/teamchallenge/bookti/review/UserReviewServiceImp.java]:
```suggestion
        if (Objects.equals(reviewerId, review.getOwnerId())) {
```
Or remove a review owner from DTO and extract owner id from access token

File: src/main/java/com/teamchallenge/bookti/review/UserReviewServiceImp.java [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+            throw new UserNotFoundException("Write correct Users id.");
+        }
+    }
+
+    @Transactional
+    @Override
+    public String delete(Integer reviewId) {
+        UserReview userReviewer = userReviewRepository.findById(reviewId)
+                .orElseThrow(
+                        () -> new UserNotFoundException(format("Reviewer's User with id <{0}> not found.", reviewId)));
```
Comment (MinoUni) [inline src/main/java/com/teamchallenge/bookti/review/UserReviewServiceImp.java]: Replace with `existsById(reviewId)` method, it doesn't need to use resources to read data from db and then delete it.

File: src/test/java/com/teamchallenge/bookti/ApplicationTests.java
Comment (MinoUni) [inline src/test/java/com/teamchallenge/bookti/ApplicationTests.java:1]: Return test to check that spring context loads correctly

File: src/test/java/com/teamchallenge/bookti/review/ReviewControllerTest.java [hunk 12]
```diff
@@ -0,0 +1,338 @@
...
+                .andExpect(jsonPath("$.details[*]", containsInAnyOrder("Field <rating> must not be null",
+                        "Field <message> must not be blank",
+                        "Field <ownerId> must not be null")))
+                .andExpect(jsonPath("status_code").value(HttpStatus.BAD_REQUEST.value()));
+
+        verify(reviewService, times(0)).findAllUserReceivedReviewsById(ownerId);
+        verify(reviewService, times(0)).save(any(UserReview.class));
+    }
+
+}
```
Comment (MinoUni) [inline src/test/java/com/teamchallenge/bookti/review/ReviewControllerTest.java:358]: add whitespace

File: src/test/java/com/teamchallenge/bookti/review/UserReviewRepositoryTest.java [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+        userReviewRepository.save(userReview);
+
+        UserReview newUserReview = entityManager.find(UserReview.class, userReview.getId());
+
+        assertAll(
+                () -> assertEquals(userReview.getId(), newUserReview.getId()),
+                () -> assertEquals(userReview.getMessage(), newUserReview.getMessage())
+        );
+    }
+}
```
Comment (MinoUni) [inline src/test/java/com/teamchallenge/bookti/review/UserReviewRepositoryTest.java:119]: add whitespace

File: src/test/java/com/teamchallenge/bookti/review/UserReviewServiceImpTest.java [hunk 14]
```diff
@@ -0,0 +1,239 @@
...
+
+        String accessToken = "Bearer eyJhbGciOiJSUzI1NiJ9.eyJpc3MiOiJCb29r" +
+                "E5DfYaLtRIND1SIYtvR1y1bfVjvqxS_hCJuFgE1MRiuqF1SN1vDftrkLUUI7YUER_iqqb-aQ";
+
+        Integer result = userReviewService.getUserIdFromAccessToken(accessToken);
+
+        assertEquals(Integer.valueOf(userId), result);
+
+    }
+}
```
Comment (MinoUni) [inline src/test/java/com/teamchallenge/bookti/review/UserReviewServiceImpTest.java:229]: add whitespace

**Review by MinoUni — CHANGES_REQUESTED** [review]
1. Install google-java-formatter plugin and format code;
2. Refactor namings corresponding to requested changes(request paths, tests namings, etc);

File: pom.xml (hunk 1 above)
Comment (Lyubomy-r) [reply pom.xml]: Виправив

File: pom.xml (hunk 2 above)
Comment (Lyubomy-r) [reply pom.xml]: Виправив

File: src/main/java/com/teamchallenge/bookti/review/ReviewController.java (hunk 3 above)
Comment (Lyubomy-r) [reply src/main/java/com/teamchallenge/bookti/review/ReviewController.java]: Виправив

File: src/main/java/com/teamchallenge/bookti/review/ReviewController.java (hunk 4 above)
Comment (Lyubomy-r) [reply src/main/java/com/teamchallenge/bookti/review/ReviewController.java]: Виправив

File: src/main/java/com/teamchallenge/bookti/review/ReviewController.java (hunk 5 above)
Comment (Lyubomy-r) [reply src/main/java/com/teamchallenge/bookti/review/ReviewController.java]: Виправив

File: src/main/java/com/teamchallenge/bookti/review/UserReviewServiceImp.java (hunk 11 above)
Comment (Lyubomy-r) [reply src/main/java/com/teamchallenge/bookti/review/UserReviewServiceImp.java]: Виправив

File: src/test/java/com/teamchallenge/bookti/ApplicationTests.java
Comment (Lyubomy-r) [reply src/test/java/com/teamchallenge/bookti/ApplicationTests.java:1]: Виправив

File: src/test/java/com/teamchallenge/bookti/review/ReviewControllerTest.java (hunk 12 above)
Comment (Lyubomy-r) [reply src/test/java/com/teamchallenge/bookti/review/ReviewControllerTest.java:358]: Виправив

File: src/test/java/com/teamchallenge/bookti/review/UserReviewRepositoryTest.java (hunk 13 above)
Comment (Lyubomy-r) [reply src/test/java/com/teamchallenge/bookti/review/UserReviewRepositoryTest.java:119]: Виправив

File: src/test/java/com/teamchallenge/bookti/review/UserReviewServiceImpTest.java (hunk 14 above)
Comment (Lyubomy-r) [reply src/test/java/com/teamchallenge/bookti/review/UserReviewServiceImpTest.java:229]: Виправив

**Review by MinoUni — APPROVED** [review]

**Pull request merged**